File: yearofthecow.py
n=int(input())
years=["Ox", "Tiger", "Rabbit", "Dragon", "Snake", "Horse", "Goat", "Monkey", "Rooster", "Dog", "Pig", "Rat"]

arr=[]
for i in range(n):
	line = input().split(' ')
	arr.append([line[0], line[3], line[4], line[7]])

for i in range(n):
    
    if arr[i][3]=='Bessie':
        arr[i].append(years.index(arr[i][2]))
        arr[i].append(0)
    else:
        for j in range(i-1,-1,-1):
            if arr[j][0] == arr[i][3]:
                value = j
        arr[i].append(years.index(arr[i][2]))
        arr[i].append(years.index(arr[value][2]))
    if arr[i][1] == 'previous':
        if arr[i][4] >= arr[i][5]:
            arr[i].append(-12 + arr[i][4] - arr[i][5])
        else:
            arr[i].append(arr[i][4] - arr[i][5])
    else:
        if arr[i][4] <= arr[i][5]:
            arr[i].append(12 - (arr[i][5] - arr[i][4]))
        else:
            arr[i].append(arr[i][4] - arr[i][5])
    
    
    if arr[i][3] == 'Bessie':
        arr[i].append(arr[i][6])
    else:
        for j in range(i - 1, -1, -1):
            if arr[j][0] == arr[i][3]:
                value = j
                arr[i].append(arr[i][6] + arr[value][7])
    if arr[i][0] == 'Elsie':
        ans = arr[i][7]
        break

print(abs(ans))

File: 2021FebBronze/yearofthecow.cpp
#include <bits/stdc++.h>
using namespace std;

string animals[12] = {"Ox", "Tiger", "Rabbit", "Dragon", "Snake", "Horse", "Goat", "Monkey", "Rooster", "Dog", "Pig", "Rat"};

map<string, int> whenwasborn;

int n;

string getanimal(int year)
{
    int a = 0;
    int y = 2021;
    if (y < year)
        while (y < year)
        {
            y++;
            a++;
            if (a == 12)
                a = 0;
        }
    else
        while (y > year)
        {
            y--;
            a--;
            if (a == -1)
                a = 11;
        }
    return animals[a];
}

int main()
{
    cin >> n;
    whenwasborn["Bessie"] = 2021;
    string cowa, born, in, relation, animal, year, from, cowb;
    for (int i = 0; i < n; i++)
    {
        cin >> cowa >> born >> in >> relation >> animal >> year >> from >> cowb;
        whenwasborn[cowa] = whenwasborn[cowb];

        if (relation == "previous")
            whenwasborn[cowa]--;
        else
            whenwasborn[cowa]++;

        while (getanimal(whenwasborn[cowa]) != animal)
        {
            if (relation == "previous")
                whenwasborn[cowa]--;
            else
                whenwasborn[cowa]++;
        }
    }
    int ans = whenwasborn["Bessie"] - whenwasborn["Elsie"];
    cout << abs(ans) << endl;
}
